spiffe/src/workload_api/x509_source.rs

```diff
@@ -0,0 +1,449 @@
+            closed: RwLock::new(false),
+            svid_picker,
+            cancellation_token,
+            updated,
+            update_notifier,
+        });
+
+        let source_clone = Arc::clone(&source);
+        let mut client_clone = source_clone.workload_api_client.clone();
+        let mut stream = client_clone
+            .stream_x509_contexts()
+            .await
+            .map_err(|e| X509SourceError::GrpcError(GrpcClientError::from(e)))?;
+
+        // Block until the first X509Context is fetched.
+        if let Some(update) = stream.next().await {
+            match update {
+                Ok(x509_context) => source_clone.set_x509_context(x509_context).map_err(|e| {
+                    X509SourceError::Other(format!("Failed to set X509Context: {}", e))
+                })?,
+                Err(e) => return Err(X509SourceError::GrpcError(GrpcClientError::from(e))),
+            }
+        } else {
+            return Err(X509SourceError::Other(
+                "Stream ended without an update".to_string(),
+            ));
+        }
+
+        // Spawn a task to handle subsequent updates
+        tokio::spawn(async move {
+            loop {
+                if cancellation_token_clone.is_cancelled() {
+                    debug!("Cancellation signal received; stopping updates.");
+                    break;
+                }
+
+                match stream.next().await {
+                    Some(update) => match update {
+                        Ok(x509_context) => {
+                            if let Err(e) = source_clone.set_x509_context(x509_context) {
+                                error!("Error updating X509 context: {}", e);
+                            } else {
+                                info!("X509 context updated successfully.");
+                            }
+                        }
+                        Err(e) => error!("GRPC client error: {}", e),
+                    },
+                    None => {
+                        error!("Stream ended; no more updates will be received.");
+                        break;
+                    }
+                }
+            }
+        });
+
+        Ok(source)
+    }
+
+    fn set_x509_context(&self, x509_context: X509Context) -> Result<(), X509SourceError> {
+        let svid = if let Some(ref svid_picker) = self.svid_picker {
+            svid_picker
+                .pick_svid(&x509_context.svids())
+                .ok_or(X509SourceError::NoSuitableSvid)?
+        } else {
+            x509_context
+                .default_svid()
+                .ok_or(X509SourceError::NoSuitableSvid)?
+        };
+
+        self.set_svid(&svid)?;
+
+        self.bundles
+            .write()
+            .map_err(|e| {
+                X509SourceError::WriteLockAcquisitionFailed("bundles".to_string(), e.to_string())
+            })?
+            .replace(x509_context.bundle_set().clone());
+
+        self.notify_update();
```

Comment:
Do we want to send the context itself, or would that defeat the purpose because you could hone just subscribe to the x509_context stream directly.

Reply:
It's a good question. 

I think that sending the context itself would defeat the purpose of the `X509Source`, as it is designed to maintain the `x509_context` update in memory for users to consume. If users just want to subscribe to the `x509_context` stream directly, they wouldn't need the `X509Source`, they can use the `WorkloadApiClient`. Therefore, it makes more sense to keep the current design where the `X509Source` only signals an update, allowing users to retrieve the context through its provided methods.